homeassistant/components/iotty/api.py

```diff
@@ -0,0 +1,108 @@
+"""API for iotty bound to Home Assistant OAuth."""
+
+import asyncio
+from contextlib import suppress
+import logging
+from typing import Any
+
+from aiohttp import ClientSession
+from iottycloud.cloudapi import CloudApi
+from iottycloud.device import Device
+from iottycloud.verbs import RESULT, STATUS
+
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.entity import Entity
+
+from .const import IOTTYAPI_BASE, OAUTH2_CLIENT_ID
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
+
+
+class IottyProxy(CloudApi):
+    """Provide iotty authentication tied to an OAuth2 based config entry."""
+
+    _devices: list[Device]
+    _entities: dict
+    _coroutine: Any
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        websession: ClientSession,
+        oauth_session: config_entry_oauth2_flow.OAuth2Session,
+    ) -> None:
+        """Initialize iotty auth."""
+
+        super().__init__(websession, IOTTYAPI_BASE, OAUTH2_CLIENT_ID)
+        self._oauth_session = oauth_session
+        self._devices = []
+        self._entities = {}
+        self._hass = hass
+
+    async def init(self, entry):
+        """Initialize iotty middleware."""
+        with suppress(Exception):
+            self._devices = await self.get_devices()
+
+        _LOGGER.debug("info There are %d Devices", len(self._devices))
+
+        await self._hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+        if len(self._devices) > 0:
+            self._coroutine = self._hass.async_create_background_task(
+                self._polling(), "polling_task"
+            )
+
+    async def devices(self, device_type: str) -> Any:
+        """Get devices for a specific type."""
+
+        _LOGGER.debug("There are %d devices", len(self._devices))
+
+        ret = [d for d in self._devices if d.device_type == device_type]
+
+        return ret
+
+    def store_entity(self, device_id: str, entity: Entity) -> None:
+        """Test."""
+        _LOGGER.debug("store_entity %s", device_id)
+        self._entities[device_id] = entity
+
+    async def _polling(self) -> None:
```

Comment:
Can we use a Coordinator instead?

Reply:
We can certainly. Is there a specific reason to do so?

Reply:
Anyway, will address this in next commits

Reply:
The coordinator does the scheduling and co already and we should reuse it instead of creating all by our own. Please add a coordinator

Also if the user disable a certain device, it looks like this API is still polling the information about that device, which can be omitted by using a coordinator

Reply:
Ok, implemented coordinator